backends/magma/ceed-magma-restriction.c

```diff
@@ -34,27 +34,57 @@ static int CeedElemRestrictionApply_Magma(CeedElemRestriction r,
   CeedInt NCOMP;
   CeedElemRestrictionGetNumComponents(r, &NCOMP);
 
-  const CeedScalar *du;
-  CeedScalar *dv;
-  ierr = CeedVectorGetArrayRead(u, CEED_MEM_DEVICE, &du); CeedChk(ierr);
-  ierr = CeedVectorGetArray(v, CEED_MEM_DEVICE, &dv); CeedChk(ierr);
-
   if (!impl->indices) {  // Strided Restriction
-    CeedInt strides[3];
-    ierr = CeedElemRestrictionGetStrides(r, &strides); CeedChk(ierr);
-    CeedInt *dstrides;
-    ierr = magma_malloc( (void **)&dstrides,
-                         3 * sizeof(CeedInt)); CeedChk(ierr);
-    magma_setvector(3, sizeof(CeedInt), strides, 1, dstrides, 1);
 
-    if (tmode == CEED_TRANSPOSE) {
-      magma_writeDofsStrided(NCOMP, nnodes, esize, nelem, dstrides, du, dv);
+    // Check to see if vector is already in magma Q-Vector layout
+    //  (dimension = slowest index, then component, then element,
+    //    then node)
+    bool backendstrides;
+    ierr = CeedElemRestrictionGetBackendStridesStatus(r, &backendstrides);
+
+    if (backendstrides) {
+
+      CeedScalar *du;
+
+      // Set pointer to same data; no re-ordering needed
+      ierr = CeedVectorGetArray(u, CEED_MEM_DEVICE, &du); CeedChk(ierr);
+      ierr = CeedVectorSetArray(v, CEED_MEM_DEVICE, CEED_USE_POINTER, du);
+      ierr = CeedVectorRestoreArray(u, &du); CeedChk(ierr);
```

Comment:
I think it would be safer(?)/clearer to handle this optimization at the operator level. Applying a restriction and then modifying u should not result in a change in v, so I think alising u as v could result in bugs that might be hard for a user to trace.

The operator should instead realize that a strided restriction with the backend ordering is being used and use u in place of v. (Need to add in this PR for CUDA, but this would be a new performance improvement to the GPU backends).

For the CUDA/OCCA backends, I'm using the backend's preferred strides if this function is ever actually called with CEED_STRIDES_BACKEND.

Reply:
That makes sense.  I changed it in 98df188a45fac370d96950e2cefb88ef466f8462 to set the strides explicitly with the layout we're using for the MAGMA backend.

Reply:
Thanks